src/Features/CSharp/Portable/MakeMethodAsynchronous/CSharpMakeMethodAsynchronousCodeFixProvider.cs

```diff
@@ -54,46 +55,93 @@ protected override bool IsAsyncSupportingFunctionSyntax(SyntaxNode node)
         {
-            var newReturnType = FixMethodReturnType(keepVoid, methodSymbol, localFunction.ReturnType, taskType, taskOfTType, valueTaskOfTType);
+            var newReturnType = FixMethodReturnType(keepVoid, methodSymbol, localFunction.ReturnType, knownTypes);
             var newModifiers = AddAsyncModifierWithCorrectedTrivia(localFunction.Modifiers, ref newReturnType);
             return localFunction.WithReturnType(newReturnType).WithModifiers(newModifiers);
         }
 
         private static TypeSyntax FixMethodReturnType(
-            bool keepVoid, IMethodSymbol methodSymbol, TypeSyntax returnType,
-            INamedTypeSymbol taskType, INamedTypeSymbol taskOfTType, INamedTypeSymbol valueTaskOfTType)
+            bool keepVoid, IMethodSymbol methodSymbol, TypeSyntax returnTypeSyntax,
+            KnownTypes knownTypes)
         {
-            var newReturnType = returnType.WithAdditionalAnnotations(Formatter.Annotation);
+            var newReturnType = returnTypeSyntax.WithAdditionalAnnotations(Formatter.Annotation);
 
             if (methodSymbol.ReturnsVoid)
             {
                 if (!keepVoid)
                 {
-                    newReturnType = taskType.GenerateTypeSyntax();
+                    newReturnType = knownTypes.TaskType.GenerateTypeSyntax();
                 }
             }
             else
             {
-                if (!IsTaskLike(methodSymbol.ReturnType, taskType, taskOfTType, valueTaskOfTType))
+                var returnType = methodSymbol.ReturnType;
+                if (IsIEnumerable(returnType, knownTypes) && IsIterator(methodSymbol))
+                {
+                    newReturnType = MakeGenericType("IAsyncEnumerable", methodSymbol.ReturnType);
+                }
+                else if (IsIEnumerator(returnType, knownTypes) && IsIterator(methodSymbol))
+                {
+                    newReturnType = MakeGenericType("IAsyncEnumerator", methodSymbol.ReturnType);
+                }
+                else if (IsIAsyncEnumerableOrEnumerator(returnType, knownTypes))
+                {
+                    // Leave the return type alone
+                }
+                else if (!IsTaskLike(returnType, knownTypes))
                 {
                     // If it's not already Task-like, then wrap the existing return type
                     // in Task<>.
-                    newReturnType = taskOfTType.Construct(methodSymbol.ReturnType).GenerateTypeSyntax();
+                    newReturnType = knownTypes.TaskOfTType.Construct(methodSymbol.ReturnType).GenerateTypeSyntax();
                 }
             }
 
-            return newReturnType.WithTriviaFrom(returnType);
+            return newReturnType.WithTriviaFrom(returnTypeSyntax);
+
+            TypeSyntax MakeGenericType(string type, ITypeSymbol typeArgumentFrom)
+            {
+                return SyntaxFactory.GenericName(
+                    SyntaxFactory.Identifier(type),
```

Comment:
this seems wrong.  how will this work if they don't have an import for the namespace where these types live?  We should be fully qualifying these.

Reply:
In that case you will get an error and a suggestion to import the namespace.
I'm ok to fully qualify though.

Reply:
It's easy to do that too just using SyntaxGenerator.TypeExpression.  Or ITypeSymbol.GenerateTypeSyntax.  They will even put the right annotations on things so things will simplify properly :)